onnxruntime/python/onnxruntime_pybind_mlvalue.cc

```diff
@@ -426,11 +426,27 @@ static void CopyDataToTensor(PyArrayObject* darray, int npy_type, Tensor& tensor
 
 inline void CopyDataToTensor(PyArrayObject* darray, int npy_type, std::unique_ptr<Tensor>& p_tensor,
                              MemCpyFunc mem_cpy_to_device = CpuToCpuMemCpy) {
-  CopyDataToTensor(darray, npy_type, *p_tensor, mem_cpy_to_device);
+  // If the provided py_array is already contiguous, its ref count will be incremented and we will decrement
+  // the ref count when this function's scope ends
+  PyArrayObject* contiguous_array = PyArray_GETCONTIGUOUS(darray);
+  ORT_ENFORCE(contiguous_array != nullptr, "The array that needs to be copied over to the tensor must be contiguous");
+
+  // Clean-up the contiguous array after the end of the current scope
+  UniqueDecRefPtr<PyArrayObject> contiguous_array_guard(contiguous_array, DecRefFn<PyArrayObject>());
+
+  CopyDataToTensor(contiguous_array, npy_type, *p_tensor, mem_cpy_to_device);
 }
 
 void CopyDataToTensor(const py::array& py_array, int npy_type, Tensor& tensor, MemCpyFunc mem_cpy_to_device) {
-  CopyDataToTensor(reinterpret_cast<PyArrayObject*>(py_array.ptr()), npy_type, tensor, mem_cpy_to_device);
+  // If the provided py_array is already contiguous, its ref count will be incremented and we will decrement
```

Comment:
I think this function can be simplyfied. The py::array object already holds the reference. We can check if the array is contigious by calling a method and then simply call data() to obtain a pointer. There is an example of that.

The above function can also be simplified by wrapping darray into py::array

Reply:
Not quite following the first point - I would have thought `PyArray_GETCONTIGUOUS()` does exactly what you are talking about and we simply decrement the ref count at the end (because we call PyArray_GETCONTIGUOUS). Is there anything simpler than just these 2 lines of code ? Can you reference the example you are referring to ?